src/relay/ir/expr_functor.cc

```diff
@@ -476,41 +476,31 @@ Expr Bind(const Expr& expr, const tvm::Map<Var, Expr>& args_map) {
   if (const FunctionNode* func = expr.as<FunctionNode>()) {
     Expr new_body = ExprBinder(args_map).VisitExpr(func->body);
     Array<Var> new_params;
-    std::vector<VirtualDevice> new_param_virtual_devices;
+    bool params_unchanged = true;
     for (size_t i = 0; i < func->params.size(); ++i) {
       if (!args_map.count(func->params[i])) {
         new_params.push_back(func->params[i]);
-        new_param_virtual_devices.push_back(GetFunctionParamVirtualDevice(func, i));
+      } else if (const auto var = args_map[func->params[i]].as<VarNode>()) {
+        // If we're mapping a variable to a variable and not a normal expr, then we want to
+        // put the substitution in the new parameters.
+        params_unchanged = false;
+        new_params.push_back(GetRef<Var>(var));
       }
     }
-    if (new_body.same_as(func->body) && new_params.size() == func->params.size()) {
+    if (new_body.same_as(func->body) && new_params.size() == func->params.size() &&
+        params_unchanged) {
       return expr;
     }
     auto ret =
         Function(new_params, new_body, func->ret_type, func->type_params, func->attrs, func->span);
-    ret =
-        MaybeFunctionOnDevice(ret, new_param_virtual_devices, GetFunctionResultVirtualDevice(func));
-    std::unordered_set<Var, ObjectPtrHash, ObjectPtrEqual> set;
```

Comment:
Bind should really be called SubstAndBind, since it assumes any free vars introduced in the subst rhs should be implicitly bound in the result function. I think you'll need to keep this semantics.

For your purposes I think just directly using the ExprBinder is what you want since you're simply wanting the Subst aspect. I forgot about this hackery when I pointed you to this func.

Reply:
Yeah I think that makes more sense, I'll revert the changes in Bind and use ExprBinder instead